### ISSUE TITLE
[BUG] ariaTSsetup.py with -mo option fails

### ISSUE THREAD

**Issue by EJFielding** [aria-tools/ARIA-tools#188]
**Describe the bug**

I added the `-mo 15000` option to my `ariaTSsetup.py` run to select only the pairs with minimumOverlap of greater than 15000 km^2. It worked for one bbox, but not for another.

**To Reproduce**
1. Download 
`ariaDownload.py --bbox "26 28 83 86" --track 19 -o download`
2. Command used
```
# unit of minimumOverlap are km^2
ariaTSsetup.py -f 'products/*.nc' --bbox "26 28 83 86" -d download --mask download -w bbox26_28 -mo 15000
```
3. Error Output
```
***Time-series Preparation Function:***
Shapefile bbox26_28/user_bbox.json created for input user bounds.
Multi-core version
Multi-core version failed, will try single for loop
285 out of 470 GUNW products rejected for not meeting user's bbox spatial criteria
Group GUNW products into spatiotemporally continuous interferograms.
All (163) interferograms are spatially continuous.
Thread count specified for gdal multiprocessing = 2
Traceback (most recent call last):
  File "/Users/fielding/anaconda3/envs/aria-tools/bin/ariaTSsetup.py", line 15, in <module>
    main(inps)
  File "/Users/fielding/anaconda3/envs/ARIA-tools/lib/python3.8/site-packages/ARIAtools/tsSetup.py", line 254, in main
    standardproduct_info.products[0], standardproduct_info.products[1], standardproduct_info.bbox_file, prods_TOTbbox, prods_TOTbbox_metadatalyr, arrshape, proj = merged_productbbox(standardproduct_info.products[0], standardproduct_info.products[1], os.path.join(inps.workdir,'productBoundingBox'), standardproduct_info.bbox_file, inps.croptounion, num_threads=inps.num_threads, minimumOverlap=inps.minimumOverlap, verbose=inps.verbose)
  File "/Users/fielding/anaconda3/envs/ARIA-tools/lib/python3.8/site-packages/ARIAtools/extractProduct.py", line 226, in merged_productbbox
    overlap_area=shapefile_area(prods_bbox)
  File "/Users/fielding/anaconda3/envs/ARIA-tools/lib/python3.8/site-packages/ARIAtools/shapefile_util.py", line 77, in shapefile_area
    lon, lat=file_bbox.exterior.coords.xy
AttributeError: 'MultiPolygon' object has no attribute 'exterior'
```
**Expected behavior**

This worked with a slightly smaller bbox `ariaTSsetup.py -f 'products/*.nc' --bbox "26 27.5 83 86" -d download --mask download -w bbox26_27_5 -mo 15000` and found a total of 11 pairs that had adequate coverage (see https://github.com/aria-tools/ARIA-tools-docs/issues/35). It should also work with the bigger bbox.

**Screenshots**

**Desktop (please complete the following information):**
 - ARIA-tools git tag: [e.g. git show --summary]
```
commit 7d173b0aa2b4a3a6459ac674f253c7e4058b803f (HEAD -> dev, origin/dev, origin/HEAD)
Author: rzinke <48299159+rzinke@users.noreply.github.com>
Date:   Thu Jul 9 18:17:29 2020 -0700
```
 - OS
MacOS

**Additional context**

Present coverage of Nepal is only partial due to some issues with transferring processed data from ARIA system to ASF.

**Comment by sssangha** [comment]
I cannot replicate this issue. My hunch is that perhaps you have encountered a corrupted file, perhaps one that had not finished downloading completely.

To test this, can you please try to help me track down the problematic product as so:
1. add ```print(scene['unwrappedPhase'][0])``` right above line 226 in ```"/Users/fielding/anaconda3/envs/ARIA-tools/lib/python3.8/site-packages/ARIAtools/extractProduct.py"```
2. Run ```gdalinfo``` on the printed path and let me know whether or not an error is raised.

If it is indeed a corrupted product, I can implement a fix to exit with error or just skip corrupted products. I'm leaning more on implementing the former as it should be straightforward to alert users such that they may attempt to redownload and salvage products, instead of just moving on without them by default.

**Comment by rzinke** [comment]
I was able to replicate the error. 
@sssangha, I added `print(scene['unwrappedPhase'][0])` to line 226 in my `/python3.6/site-packages/ARIAtools/extractProduct.py` script. It produced this list of products:
NETCDF:"/u/sar-r2/rzinke/ARIAfix/OlapTest/products/S1-GUNW-D-R-019-tops-20150511_20150429-001856-29770N_27949N-PP-8465-v2_0_3.nc":/science/grids/data/unwrappedPhase
NETCDF:"/u/sar-r2/rzinke/ARIAfix/OlapTest/products/S1-GUNW-D-R-019-tops-20150523_20150429-001905-29441N_27449N-PP-c099-v2_0_3.nc":/science/grids/data/unwrappedPhase
NETCDF:"/u/sar-r2/rzinke/ARIAfix/OlapTest/products/S1-GUNW-D-R-019-tops-20150523_20150511-001905-29441N_27449N-PP-1cca-v2_0_3.nc":/science/grids/data/unwrappedPhase
NETCDF:"/u/sar-r2/rzinke/ARIAfix/OlapTest/products/S1-GUNW-D-R-019-tops-20150604_20150511-001858-29770N_27950N-PP-bd89-v2_0_3.nc":/science/grids/data/unwrappedPhase
NETCDF:"/u/sar-r2/rzinke/ARIAfix/OlapTest/products/S1-GUNW-D-R-019-tops-20150604_20150523-001858-29770N_27950N-PP-68db-v2_0_3.nc":/science/grids/data/unwrappedPhase
NETCDF:"/u/sar-r2/rzinke/ARIAfix/OlapTest/products/S1-GUNW-D-R-019-tops-20150616_20150523-001901-29774N_27781N-PP-1c63-v2_0_3.nc":/science/grids/data/unwrappedPhase

I ran gdalinfo on the last one in the list, and no error was raised. I tried removing that product from the products folder, then deleted the bbox26_28 folder, then reran the same command. It produced the same error.

**Comment by rzinke** [comment]
Here is the error I get when I remove all the above ifgs from the products folder:

(ARIA-tools) [rzinke@sar OlapTest]$ ariaTSsetup.py -f 'products/*.nc' --bbox "26 28 83 86" -d download --mask download -w bbox26_28 -mo 15000
***Time-series Preparation Function:***
Shapefile bbox26_28/user_bbox.json created for input user bounds.
Multi-core version
All (179) GUNW products meet spatial bbox criteria.
Group GUNW products into spatiotemporally continuous interferograms.
All (157) interferograms are spatially continuous.
Thread count specified for gdal multiprocessing = 2
NETCDF:"/u/sar-r2/rzinke/ARIAfix/OlapTest/products/S1-GUNW-D-R-019-tops-20150616_20150604-001901-29774N_27781N-PP-5d92-v2_0_3.nc":/science/grids/data/unwrappedPhase
Traceback (most recent call last):
  File "/u/sar-r0/rzinke/python/miniconda3/envs/ARIA-tools/bin/ariaTSsetup.py", line 15, in <module>
    main(inps)
  File "/u/sar-r0/rzinke/python/miniconda3/envs/ARIA-tools/lib/python3.6/site-packages/ARIAtools/tsSetup.py", line 254, in main
    standardproduct_info.products[0], standardproduct_info.products[1], standardproduct_info.bbox_file, prods_TOTbbox, prods_TOTbbox_metadatalyr, arrshape, proj = merged_productbbox(standardproduct_info.products[0], standardproduct_info.products[1], os.path.join(inps.workdir,'productBoundingBox'), standardproduct_info.bbox_file, inps.croptounion, num_threads=inps.num_threads, minimumOverlap=inps.minimumOverlap, verbose=inps.verbose)
  File "/u/sar-r0/rzinke/python/miniconda3/envs/ARIA-tools/lib/python3.6/site-packages/ARIAtools/extractProduct.py", line 227, in merged_productbbox
    overlap_area=shapefile_area(prods_bbox)
  File "/u/sar-r0/rzinke/python/miniconda3/envs/ARIA-tools/lib/python3.6/site-packages/ARIAtools/shapefile_util.py", line 77, in shapefile_area
    lon, lat=file_bbox.exterior.coords.xy
AttributeError: 'MultiPolygon' object has no attribute 'exterior'

**Comment by EJFielding** [comment]
@rzinke Thanks for trying this on the Sar Linux machine. It looks like exactly the same error I was getting on my Mac.

**Comment by dbekaert** [comment]
@rzinke @sssangha  More specifically upon merging of products to individual contiguous IFGS, area is being calculated. Products not continuous should be rejected before. 

Although should be rejected for some reason this seems to be a multi-polygon. Although did not yet see why:
```
MULTIPOLYGON (((85.8589054789707 27.61616635835893, 84.9868923614458 27.7632830196431, 85.00975375166951 27.8712205799762, 84.21405315208079 28, 85.9431834926482 28, 85.8589054789707 27.61616635835893)), ((84.16400526663317 28, 84.15375175956071 27.9496219247077, 83.82994054141908 28, 84.16400526663317 28)))
```

Generally the products merged are a single polygon product:
```
POLYGON ((84.92069166676448 27.25786352956226, 84.9203335136462 27.257924292377, 84.92020138559693 27.25729927783146, 84.92010179636679 27.2573161733022, 84.9080404753638 27.2002614580581, 83.9849807440597 27.3509826807456, 84.0067587033119 27.4585115923867, 84.00655432990639 27.4585436333665, 84.00660189904551 27.4587785395215, 84.00623123533229 27.4588366480897, 84.0062803518445 27.459079211709, 83.2365900621936 27.579734158608, 83.3186261976778 28, 85.9471284777497 28, 85.753225566811 27.1170531018169, 84.92208837792801 27.2580627302451, 84.92203992640729 27.257833540904, 84.9219973028374 27.2578407730095, 84.9219932970871 27.2578218284958, 84.9207282214071 27.2580364494459, 84.92069166676448 27.25786352956226))
```

**Comment by sssangha** [comment]
@dbekaert @EJFielding ok I figured out the problem. I wasn't able to replicate the issue before as I had to re-run the download to access the IFGs from 2015 that raised these errors.

There isn't anything wrong with these products per se, but when the program attempts to intersect its bounding box with the user-defined box then it fails because a multi-polygon is returned. You can see why this is the case in the screenshot below, as an intersection along the extreme end of the product produces multiple IFGs:
<img width="552" alt="Screen Shot 2020-07-15 at 1 08 34 AM" src="https://user-images.githubusercontent.com/13227405/87520674-26d23500-c638-11ea-9ef9-6027d80f3848.png">

So it makes sense how changing the bbox would allow you to circumvent this "issue".

Will think about a way to handle this. This particular scene would obviously be rejected given the specified minimum threshold, but need to handle such cases appropriately in other scenarios (e.g. no minimum threshold defined and/or no bounding box defined, such that common intersection for all products is passed).

**Comment by sssangha** [comment]
To add to my previous comment, I figured out a working solution! I simply had to update the ```shapefile_area``` function under the ```shapefile_util.py``` script to handle such cases. Will issue PR soon.

@EJFielding It looks like your specified minimum threshold in fact may be a bit too conservative as it passes through short scenes:
<img width="1592" alt="Screen Shot 2020-07-15 at 1 51 17 AM" src="https://user-images.githubusercontent.com/13227405/87525178-fe4d3980-c63d-11ea-975d-36fedc271bf6.png">

**Comment by dbekaert** [comment]
@sssangha yep looks good; just need to loop over the multi-polygons.

**Comment by EJFielding** [comment]
@sssangha Thanks for tracking down the multipolygon issue. I can see that it is literally an edge case where the edge of the bounding box intersected the steps in the S1 TOPS scene boundary.

I just made a wild guess for my `-mo 15000` threshold, but I wonder if the edge case would still happen using a larger minimumOverlap. Sooner or later, someone is going to choose another `bbox` that intersects a frame boundary.

**Comment by dbekaert** [comment]
@EJFielding The bug is on the area calculation and not related to the threshold that the user selects. rather invoking the -mo option triggers the area calculation workflow. The area once calculated is compared to the threshold value.

Edge case as rare do occur. Thanks for identifying @EJFielding! This should be resolved with the PR merged in dev.

**Comment by EJFielding** [comment]
I updated to the new dev version and verified that it does work correctly for my original parameters.
```
ariaTSsetup.py -f 'products/*.nc' --bbox "26 28 83 86" -d download --mask download -w bbox26_28 -mo 15000
```
Thanks @sssangha !

I noticed that using this bbox and the small `-mo 15000` threshold picks up a lot of scenes with small coverage, so I changed it to `-mo 42000` as @sssangha suggested. Now I get only the 11 pairs with full coverage of the area. Thanks again.
